### PR TITLE
PADV-409: Add CCX Compatibility

### PR DESCRIPTION
**Pull request by kuipumu**
## Ticket

https://agile-jira.pearson.com/browse/PADV-409

## Description

This PR adds a fix that allows the LTI consumer XBlock to work on CCXs with LTI 1.3 launches.
The LTI consumer XBlock creates a model instance of LtiConfiguration using its location, on CCXs this creates a new LtiConfiguration with a client_id that is not known to the tool provider. This fixes the LTI consumer on CCXs by using the master course counterpart XBlock location with the to_block_locator method of CCXBlockLocator to retrieve the master course LtiConfiguration models instance.

## Type of Change

- [x] Modify _get_or_create_local_lti_config API function.
- [x] Add test_retrieve_ccx_config test to TestGetOrCreateLocalLtiConfiguration class.
- [x] Modify get_lti_1p3_launch_data function on LtiConsumerXBlock class.

## Testing:

- Enable CCX on LMS and Studio.
- Run the LMS: `make dev.up.lms`.
- Run studio: `make dev.up.studio`.
- Run the frontend-app-learning mfe: `make dev.up.frontend-app-learning`.
- Install `xblock-lti-consumer` on the LMS and studio.
- Enable CCX on course advance settings.
- Add 'lti_consumer' to course advance setting 'Advanced Module List'.
- Add LTI consumer XBlock to the course and set up LTI 1.3 launch parameters:

```
	LTI Version: LTI 1.3
	Tool Launch URL: https://lti.tool/launch
	Tool Initiate Login URL: https://lti.tool/login
	Tool Public Key Mode: Keyset URL
	Tool Keyset URL: https://lti.tool/jwks
```

- Add a CCX coach to the course and create a CCX.
- Go to the CCX live course and execute an LTI 1.3 launch.
- The launch should work without issues.

## Reviewers

- [x] @Squirrel18 
- [x] @Jacatove

### CODE REVIEW

File: lti_consumer/api.py [hunk 1]
```diff
@@ -34,6 +34,10 @@ def _get_or_create_local_lti_config(lti_version, block_location,
     the XBlock to be the source of truth for the LTI version, which is a user-centric perspective we've adopted.
     This allows XBlock users to update the LTI version without needing to update the database.
     """
+    # Replace CCX block locator with block locator to allow CCX launch to work.
```
Comment (Jacatove) [inline lti_consumer/api.py]: CCX Launch works if tool is configured properly. Could you please explain why now it works in terms of our context. I understand it does because the tool is only configured for the Master Course, otherwise configuring it for every new CCX only does not scale, that correct?
```suggestion
    # Replace CCX block locator with Master Course block locator...
```

File: lti_consumer/tests/unit/test_api.py [hunk 2]
```diff
@@ -175,6 +175,36 @@ def test_retrieve_existing(self):
...
+        )
+
+        # Check CCX attribute was inspected on location.
+        mock_hasattr.assert_called_once_with(ccx_location, 'ccx')
+
+        # Check CCX block locator was trnasformed to block locator.
+        ccx_location.to_block_locator.assert_called_once_with()
+
+        # Check CCX lti_config is equal to master course lti_config.
+        self.assertEqual(lti_config_retrieved, lti_config)
```
Comment (Jacatove) [inline lti_consumer/tests/unit/test_api.py]:
```suggestion
        # Check CCX attribute was inspected on location.
        mock_hasattr.assert_called_once_with(ccx_location, 'ccx')
        # Check CCX block locator was transformed to block locator.
        ccx_location.to_block_locator.assert_called_once_with()
        # Check CCX lti_config is equal to master course lti_config.
        self.assertEqual(lti_config_retrieved, lti_config)
```

File: lti_consumer/tests/unit/test_api.py [hunk 3]
```diff
@@ -175,6 +175,36 @@ def test_retrieve_existing(self):
...
+        """
+        Check if the API retrieves a model using the master course location of a CCX location.
+        """
+        master_location = 'block-v1:course+test+2020+type@problem+block@test'
+        ccx_location = Mock(return_value='ccx-block-v1:course+test+2020+type@problem+block@test')
+        ccx_location.to_block_locator.return_value = master_location
+        mock_hasattr.return_value = True
+
+        # Create master course location lti_config.
+        lti_config = LtiConfiguration.objects.create(
```
Comment (Jacatove) [inline lti_consumer/tests/unit/test_api.py:190]: What would happen at the edge case where there is no Master course lti_config and when the xblock consumer is used for a CCX the first time? Is this case even possible to exist?. @kuipumu

**Review by Jacatove — CHANGES_REQUESTED** [review]

File: lti_consumer/api.py [hunk 4]
```diff
@@ -34,6 +34,10 @@ def _get_or_create_local_lti_config(lti_version, block_location,
...
+    if hasattr(block_location, 'ccx'):
```
Comment (Squirrel18) [inline lti_consumer/api.py]: What about using the CCXLocator to check if it is a CCX: https://github.com/openedx/ccx-keys? @kuipumu

File: lti_consumer/tests/unit/test_api.py (hunk 3 above)
Comment (kuipumu) [reply lti_consumer/tests/unit/test_api.py:190]: @Jacatove There is always a master course lti_config, and even if there was such scenario there always be a lti_config created by the _get_or_create_local_lti_config

File: lti_consumer/api.py (hunk 1 above)
Comment (kuipumu) [reply lti_consumer/api.py]: @Jacatove It now works because there is no LtiConfiguration being created for the CCX XBlock, each new LtiConfiguration for LTI 1.3 has a different client_id, which is not recognized by the tool provider, since CCXs are always a copy of the master course, we can make all LTI consumer XBlocks on the CCX course by using the master course counterpart of that block location.

File: lti_consumer/tests/unit/test_api.py [hunk 5]
```diff
@@ -175,6 +176,34 @@ def test_retrieve_existing(self):
...
+        Check if the API retrieves a model using the master course location of a CCX location.
+        """
+        master_location = 'block-v1:course+test+2020+type@problem+block@test'
+        ccx_location = Mock(return_value='ccx-block-v1:course+test+2020+type@problem+block@test')
+        ccx_location.to_block_locator.return_value = master_location
+        mock_isinstance.return_value = True
+
+        # Create master course location lti_config.
+        lti_config = LtiConfiguration.objects.create(
+            location=master_location
```
Comment (Jacatove) [inline lti_consumer/tests/unit/test_api.py]:
```suggestion
            location=master_location,
```

File: lti_consumer/tests/unit/test_api.py [hunk 6]
```diff
@@ -175,6 +176,34 @@ def test_retrieve_existing(self):
...
+        )
+
+        # Get CCX location lti_config.
+        lti_config_retrieved = _get_or_create_local_lti_config(
+            lti_version="",
+            block_location=ccx_location
```
Comment (Jacatove) [inline lti_consumer/tests/unit/test_api.py]:
```suggestion
            block_location=ccx_location,
```

File: lti_consumer/api.py [hunk 7]
```diff
@@ -34,6 +35,10 @@ def _get_or_create_local_lti_config(lti_version, block_location,
     the XBlock to be the source of truth for the LTI version, which is a user-centric perspective we've adopted.
     This allows XBlock users to update the LTI version without needing to update the database.
     """
+    # Replace CCX block locator with master course block locator to allow CCX launch to work.
```
Comment (Jacatove) [inline lti_consumer/api.py:38]:
```suggestion
    # Replace CCX block locator with master course block locator to allow CCX launch to work.
    # The reason behind the change is to avoid creating a new LtiConfiguration for CCXs and reuse Master's for which its client_id is already recognized by the tool provider.
    
    
```

**Review by Jacatove — CHANGES_REQUESTED** [review]

**Review by Jacatove — APPROVED** [review]

**Review by Squirrel18 — APPROVED** [review]
